0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+- `get_register_value(idx, buffer_v)`: copy the register number `idx` in the variable `buffer_v`
+- `set_halted(exit_code)`: constraint halting the program with the exit code `exit_code`.
+
+At the beginning of the instruction, the syscall number is in the register `v0`.
+
+The list of supported syscalls are:
+- `sysMmap`, mapped to value `4090`
+- `sysBrk`, mapped to value `4045`
+- `sysClone`, mapped to value `4120`
+- `sysExitGroup`, mapped to value `4246`
+- `sysRead`, mapped to value `4003`
+- `sysWrite`, mapped to value `4004`
+- `sysFcntl`, mapped to value `4055`
+
+Except in the case of the syscall `sysExitGroup`, at the end of the instruction,
+the instruction pointer must be increased by a word (32bits) as the next
+instruction pointer.
+
+For any other syscall value, the value `0` must be set into the registers `v0`
+and `a3`.
+
+An important note is that IO operations are truncated to `4` bytes, meaning that
+the register `v0` is at most `4`.
+Another important note is that addresses must be aligned at 4 bytes. IO syscalls
+**must** support unaligned addresses, i.e. the address `42` can be given even
+though it is not a multiple of `4`.
+Computing the effective address of an address `addr` can be done using the
+bitwise AND operator with `0xFFffFFfc`, which sets the latest two bits to `0`,
+i.e. takes the previous multiple of `4` which represents the aligned memory
+address. For instance, for the unaligned address `42`, the address `40` must be
+used while fetching the memory.
+
+Note that as the syscall instruction must be encoded in a circuit and its
+behavior depends on the value in the register `v0`, we must represent its effect
+using a polynomial constraining the value `v0`. We can easily do that using a
+vanishing polynomial on the values different than the one expected.
+For instance, let's suppose `v0` is `sysFnctl` and `sysFnctl` behavior is represented
+by the polynomial `P(v0, a0, a1, a3)`, we will use the polynomial:
+```
+P_syscall_fnctl(v0, a0, a1, a3) = \
+  (v0 - 4090) *
+  (v0 - 4045) *
+  (v0 - 4120) *
+  (v0 - 4246) *
+  (v0 - 4003) *
+  (v0 - 4004) *
+  P(v0, a0, a1, a3)
+```
+
+to activate the `P_syscall_fnctl` behavior when `v0 = 4055` and deactivate it
+when it has a different value. Reduction to polynomials of at most degree `2` can be
+done using [quadracization](./0008-kimchi-polynomial-quadracization.md).
+
+We suppose quadracization is going to be used, and therefore, we will not talk
+about the degree of the different polynomial constraints, and the detailed
+description will stay high level using the API described above, with some
+additional tweaks if necessary for a specific syscall behavior.
+
+Regarding the pre-oracle data, it will be loaded as part of the witness given to
+the prover.
+In the context of generating a proof using the zkVM, the program will first be
+run outside of the zkVM, and therefore we know which data and how much data will
```

Comment:
Okay this is key. So this means, there is an interpreter of this program that extracts all needed data for this particular instance of the op program being proven validity about. So the code will generate some kind of struct storing all needed values for registers, instructions, etc, after "executing" that code, so that the circuit can be built accordingly?

If so, it seems like that execution will also give us the desired length of the keccak preimage, so the gadget can be called with the right amount and set the padding coefficients accordingly in the circuit. This seems like a good thing!